### DISCUSSION TITLE
Proposal: Allow short circuited ternary operators.

### DISCUSSION THREAD

**Discussion by scottgal** [dotnet/csharplang#1480] [Language Ideas]
Apologies if there's some known format I'm meant to follow. But longtime listener first-time caller etc...
In short, my proposal is a modification to ternary operators. This would be especially useful in Razor in ASP.NET also more generally.
So when you have for example 
`@(Model.SelectedReportIndex == file.Index ? "active in" : "")`
Which is a fairly common pattern in Razor markup for conditionally adding CSS classes inline to some HTML tag. You always wind up having to specify an empty string as the 'false' condition. What I'd really want to do is:
`@(Model.SelectedReportIndex == file.Index ? "active in")`
Hence the ternary operator is 'short-circuited' with the second action essentially being 'void' (or really the default value for the type of the true condition).
A further example may be inline in normal C# code...again the benefit is added brevity:
` IsNullOrEmpty(argumentName) ? throw new ArgumentNullException(nameof(argumentName));`
Of course this would be equivalent to:
`if (IsNullOrEmpty(argumentName))  throw new ArgumentNullException(nameof(argumentName));`
But somewhat shorter (I fully admit it's less of a compelling case in non-razor code but still a pet peeve that I need the if() 😀

**Comment by scottgal** [comment]
Confused individual. What's confusing...ternary operators or something I've written. I *know* it's only saving four characters...but my current project has several hundred of these in Razor markup.

**Comment by HaloFour** [comment]
Allowing omission of the colon and false expression would make conditional operators ambiguous to parse.  Also, the default value of string is not an empty string.

**Comment by scottgal** [comment]
Also, the default value of string is not an empty string. 
Yes I know which is why I wrote:
"(or really the default value for the type of the true condition)."

**Comment by SIkebe** [comment]
Seems ambiguous to me. Is there any language that has this sort of short circuited ternary operator?

**Comment by scottgal** [comment]
Yes although it's normally `?:`  in other languages a corollary to the binary operator `?.` which we already have. 
To answer the 'ambiguity' question...maybe the solution is `?:`? This makes it explicit that we do not require the false condition? 

https://en.wikipedia.org/wiki/Elvis_operator

**Comment by HaloFour** [comment]
The Elvis operator is quite different in that it will use the second operand if the first operand is not true, otherwise it will use the first operand.  That doesn't fit the use case you outlined above in your proposal, nor would it fit in C# since it doesn't consider anything to be "truthy" other than Boolean expressions.

The C# equivalent would be `??`, as the Elvis operator is generally intended to resolve alternate values for `null` anyway (as `null` is "false" is those languages).

**Comment by scottgal** [comment]
Ok so you've now corrected two issues which aren't related to the core proposal itself, Well done aren't you clever! Any comment on *what I actually proposed*? Or are you here to just knock down people proposing things for the first time?

**Comment by HaloFour** [comment]
@scottgal 

Clarity is important.  Pretty much every proposal here goes through the ringer, including those brought up by the team members themselves.  Don't take it personally.

**Comment by amis92** [comment]
@scottgal I think it's a good use case for creating your own operator overload. Just choose one that's fitting, in my example it's `&` because it has the same precedence as ternary conditional:

```csharp
public static string operator &(bool condition, string valueIfTrue) => condition ? valueIfTrue : null;
```

Of course, you'd have to define this operator for every type you'd want to use it with, but since you've mentioned  it's essentially one use case repeating multiple times, I think it'd work.

I just feel like it's not enough of a feature in itself, as it's pretty easily achieved by adding some code on your own :)

**Comment by svick** [comment]
@amis92 You can't add a custom operator between two framework types.

**Comment by bondsbw** [comment]
Perhaps this would work just as well (assuming an appropriate `@using static` declaration):

```c#
@If(Model.SelectedReportIndex == file.Index, "active in")
```

```c#
public static string If(bool condition, string trueValue) => condition ? trueValue : string.Empty;
```

**Comment by MisinformedDNA** [comment]
Note that `IF` [already has a meaning](https://docs.microsoft.com/en-us/dotnet/visual-basic/language-reference/operators/if-operator#if-operator-called-with-two-arguments) in VB.NET. With two parameters, `IF(A, B)` uses uses `IF(object, object)` and is equivalent to `A ?? B`. I suppose a new `IF(bool, object)` could work... I, personally, don't like "global methods" in VB.NET, so I'd hate to see them in C#.

Also, as aready stated (and naming aside) `A ?: B` in wikipedia represents `A ?? B` in C#. So that leaves `?:` available for use, right. Given we have `A ? B : C`, `A ?: B` gives similar syntax and IMO offers a reasonable solution.

So (just summarizing) code that would have been written like this

`A == null ? null : B`

could be simplified to

`A != null ?: B`

which would expand in the compiler to 

`A != null ? B : default(BType)`.
